### PR TITLE
Promote StartupProbe to beta for 1.18

### PR DESCRIPTION
**Pull request by matthyx**
Promotes the StartupProbe feature (health check) to beta for the 1.18 release.
Relates to https://github.com/kubernetes/enhancements/issues/950

```release-note
Promote StartupProbe to beta for 1.18 release
```

### CODE REVIEW

**Comment by odinuge** [comment]
Hi @matthyx!

Thanks for your work on the StartupProbe. The e2e tests for StartupProbe are failing atm., so it would be nice if you could take a look at the [issue](https://github.com/kubernetes/kubernetes/issues/82747). There is also a proposed [PR](https://github.com/kubernetes/kubernetes/pull/82748).

**Comment by matthyx** [comment]
> There is also a proposed [PR](https://github.com/kubernetes/kubernetes/pull/82748).

Well, since I want to move it out of alpha, the listed PR could be merged just after this one?
@liggitt what's your opinion?

**Comment by liggitt** [comment]
> @liggitt what's your opinion?

I don't have a lot of context on it, but these are the types of questions we'd need to answer:
* is there feedback on use of the feature?
* is there evidence there are no scale issues introduced by it?

**Comment by matthyx** [comment]
I plan to use it on a small cluster, once my admins find how to add the flag with Juju...

For the scale, don't we have some tests from sig-scaling?

**Comment by liggitt** [comment]
> For the scale, don't we have some tests from sig-scaling?

we have scale tests, but they would need to be updated (or new tests added) to include use of startup probe

**Comment by matthyx** [comment]
> to include use of startup probe

ok, I can add these to my branch, and then `prow` will run them to validate the PR?

**Comment by liggitt** [comment]
> I can add these to my branch, and then `prow` will run them to validate the PR?

I'd recommend checking with @kubernetes/sig-scalability-pr-reviews. I don't think all scale tests are available as presubmits

**Comment by matthyx** [comment]
Dear scalability team, which tests should I modify to include a startupProbe into scalability assessments?
/assign @kubernetes/sig-scalability-pr-reviews

**Comment by odinuge** [comment]
I think we should hold this one until we have passing tests and have verified that StartupProbe works as expected.

ref. https://github.com/kubernetes/kubernetes/issues/82747, one failure is related to #84178 and another one is related to #84179. I also think we should wait until we run tests in CI to make sure we avoid regressions.

/hold

**Comment by matthyx** [comment]
Thanks @odinuge I will have a look...

**Comment by matthyx** [comment]
> wait until we run tests in CI to make sure we avoid regressions

I agree, but I have already tried to ping people here:
https://github.com/kubernetes/kubernetes/pull/83570#issuecomment-539877221

**Comment by odinuge** [comment]
> I agree, but I have already tried to ping people here:
#83570 (comment)

I know. Looks like the approvers in sig-node are quite busy atm., so they are kinda slow to respond... :/ I have made a PR for a new ci task `ci-kubelet-serial-alpha`
https://github.com/kubernetes/test-infra/pull/14879 for features like this one. Will update the PR tomorrow, so we hopefully can get it in quickly. It is still possible to run the tests manually tho.

There is still plenty of time to code-freeze for `1.17`, so I think it should be possible to get it fixed before then. :smile:

**Comment by odinuge** [comment]
We now have test successfully running here: https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial-alpha :tada: 

Tests are still failing, but should be fixed in #84279.

**Comment by matthyx** [comment]
@liggitt this PR and promotion to beta is at risk because #84291 #84279 and #84179 are waiting for approval... I need your guidance, pls!

**Comment by liggitt** [comment]
I would advise reaching out to individual sig-node reviewers/approvers to determine criticality of those blocking issues (at first glance, some seem severe). If those are not able to be resolved before freeze, then promotion will likely have to wait.

**Comment by matthyx** [comment]
Well, the most critical issue is that sig-node approvers haven't got enough bandwidth. I should consider asking to become a reviewer for this part of the project.

**Comment by matthyx** [comment]
@liggitt now tests are green!

Actually they don't cover all use cases (requires merging https://github.com/kubernetes/kubernetes/pull/84291) which will fail until we merge https://github.com/kubernetes/kubernetes/pull/84279 as well.

**Comment by matthyx** [comment]
> Tests are still failing, but should be fixed in #84279.

They are indeed! But we've decided to hold off moving to beta in 1.17

**Comment by odinuge** [comment]
All tests are now passing https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial-alpha :tada:

Due to a missing image it still looks like they are failing, but that should be fixed in https://github.com/kubernetes/test-infra/pull/15277.

**Comment by matthyx** [comment]
/retest
@dchen1107 can we merge this now that code freeze is over? The goal is to get enough CI signal before 1.18 and not miss the boat this time!
/cc @liggitt

**Comment by liggitt** [comment]
should the ` [NodeAlphaFeature:StartupProbe]` tag be removed from the e2e test, and the TODO resolved?

```
	/*
		These tests are located here as they require tempSetCurrentKubeletConfig to enable the feature gate for startupProbe.
		Once the feature gate has been removed, these tests should come back to test/e2e/common/container_probe.go.
	*/
```

**Comment by liggitt** [comment]
I'm also confused by the e2e tests being labeled as Conformance tests... I didn't think anything under e2e_node was considered a conformance test. cc @johnbelamaric

**Comment by liggitt** [comment]
@wojtek-t @mm4tt for scalability visibility... has this feature been evaluated for scale impact?

**Comment by matthyx** [comment]
> and the TODO resolved?

True for the tag, but we still need the flag until we exit beta (which is the reason for this location).
I will find some time tonight to update the `NodeAlphaFeature` and the `Conformance`.
Thanks a lot for the feedback :-)

**Comment by matthyx** [comment]
@dchen1107 can we merge that one to get a good CI signal before 1.18?
/priority critical-urgent
@justaugustus can you put the 1.18 milestone here?

**Comment by mm4tt** [comment]
Hey, this is Matt from SIG Scalability. Sorry for not jumping in earlier, I was OOO.

I'm fine with merging this and then assessing the scale impact by looking out our continuous scale test. It's quite early in the release cycle so we can always rollback.

That being said, we currently face some issues with our scale tests - https://github.com/kubernetes/kubernetes/issues/86082, could you wait with merging until the issue gets fixed?

**Comment by matthyx** [comment]
@dchen1107 the issue with scale tests has been closed, we can merge now !

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -486,7 +486,7 @@ const (
 	EvenPodsSpread featuregate.Feature = "EvenPodsSpread"
 
 	// owner: @matthyx
-	// alpha: v1.16
```
Comment (liggitt) [inline pkg/features/kube_features.go]: keep the alpha line and add the beta line so we can easily see which releases this went into

**Comment by liggitt** [comment]
thanks for the update, one nit on the feature godoc, then looks good to be tagged by sig-node approvers

**Comment by matthyx** [comment]
> one nit on the feature godoc

Done, thanks @liggitt !

**Comment by liggitt** [comment]
ah, forgot the API doc includes the feature state:
```
	// This is an alpha feature enabled by the StartupProbe feature flag.
	// More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#container-probes
	// +optional
	StartupProbe *Probe `json:"startupProbe,omitempty" protobuf:"bytes,22,opt,name=startupProbe"`
```

will need to update that and run `make update` to regenerate docs/openapi, etc

**Comment by matthyx** [comment]
> will need to update that to beta

@liggitt done, thanks a lot for you sharp eye!

File: api/openapi-spec/swagger.json [hunk 2]
```diff
@@ -20800,7 +20800,7 @@
   },
   "info": {
     "title": "Kubernetes",
-    "version": "v1.18.0"
+    "version": "v1.17.0"
```
Comment (liggitt) [inline api/openapi-spec/swagger.json]: revert this line, verify scripts will be unhappy (your local tags are not up-to-date, fyi)

Comment (matthyx) [reply api/openapi-spec/swagger.json]: Sorry for all this back and forth...

**Comment by liggitt** [comment]
/lgtm
/approve
for API changes

/hold for node approval, unhold when ready

**Comment by matthyx** [comment]
@liggitt can you LGTM again? I had to resolve a conflict in `swagger.json`.

**Comment by liggitt** [comment]
flake https://github.com/kubernetes/kubernetes/issues/86181

**Pull request merged**